diff --git a/kernel/selinux/selinux.c b/kernel/selinux/selinux.c
index cc440041..88c79568 100644
--- a/kernel/selinux/selinux.c
+++ b/kernel/selinux/selinux.c
@@ -99,11 +99,11 @@ struct lsm_context {
 
 static int __security_secid_to_secctx(u32 secid, struct lsm_context *cp)
 {
-    return security_secid_to_secctx(secid, (struct lsmcontext *)cp);
+    return security_secid_to_secctx(secid, &cp->context, &cp->len);
 }
 static void __security_release_secctx(struct lsm_context *cp)
 {
-    security_release_secctx((struct lsmcontext *)cp);
+    security_release_secctx(cp->context, cp->len);
 }
 #else
 #define __security_secid_to_secctx security_secid_to_secctx
